lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -241,78 +239,107 @@ private static Domain getDomain(
         }
 
         if (type.equals(BIGINT) || type.equals(INTEGER) || type.equals(DATE) || type.equals(SMALLINT) || type.equals(TINYINT)) {
-            ParquetLongStatistics longStatistics = fromNumber((Number) min, (Number) max);
-            if (isStatisticsOverflow(type, longStatistics)) {
-                return Domain.create(ValueSet.all(type), hasNullValue);
+            List<Range> ranges = new ArrayList<>();
+            for (int i = 0; i < minimums.size(); i++) {
+                long min = longExact((Number) minimums.get(i));
+                long max = longExact((Number) maximums.get(i));
+                if (isStatisticsOverflow(type, min, max)) {
+                    return Domain.create(ValueSet.all(type), hasNullValue);
+                }
+
+                ranges.add(Range.range(type, min, true, max, true));
             }
-            return createDomain(type, hasNullValue, longStatistics);
+
+            return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
         }
 
         if (type instanceof DecimalType) {
             DecimalType decimalType = (DecimalType) type;
+            List<Range> ranges = new ArrayList<>();
             if (decimalType.isShort()) {
-                ParquetLongStatistics longStatistics = min instanceof Binary && max instanceof Binary
-                        ? fromBinary(decimalType, (Binary) min, (Binary) max)
-                        : fromNumber((Number) min, (Number) max);
-                if (isStatisticsOverflow(type, longStatistics)) {
-                    return Domain.create(ValueSet.all(type), hasNullValue);
+                for (int i = 0; i < minimums.size(); i++) {
+                    Object min = minimums.get(i);
+                    Object max = maximums.get(i);
+
+                    long minValue = min instanceof Binary ? getShortDecimalValue(((Binary) min).getBytes()) : longExact((Number) min);
+                    long maxValue = min instanceof Binary ? getShortDecimalValue(((Binary) max).getBytes()) : longExact((Number) max);
```

Comment:
Going through `(Number)` is either unnecessary (eg when type is nothing else but Long), or potentially incorrect (eg should the value be BigDecimal or BigInteger, or anything else without exact long representation). We have had bugs being masked by `(Number)` usage.

I know this is prexisting, so am fine keeping for now, but we should have a TODO here.

Reply:
I think this was because the stats can be any of Integer, Long, Short, etc. Maybe that doesn't really happen, and we can just cast to Long. I'll change the code to be explicit about what types are supported.